### PR TITLE
SSO: Update name and email on each login

### LINKED ISSUE #20362: SSO login should update name and email on subsequent logins
### Describe the Improvement

We are using Keycloak as an OpenID provider and noticed that when changing a user's name or email in Keycloak, this change does not get reflected in Directus. Our users are managed via Keycloak so this should be the single source of truth regarding this information.

I have made some tweaks to `ldap.ts`, `oauth2.ts` and `openid.ts` which you can see in the attached merge request.

This is a change in behaviour that might be undesired for you, but I thought I'd raise this issue just in case someone is interested.

However, if you'd like this to be included, there are questions relating to the changes that need to be sorted out first:

- The `auth_data` prop that is currently being updated is being set with the non-falsy value or `null` which I understand to be used to delete an existing value, correct? I have not used this for the names, but rather pass the value as is (`undefined` if not existing on the user info) so that the existing name remains if it is not supplied. Please correct me, if that is not the case.
- I have only made the changes for the above-mentioned files because I did not understand how this is handled in SAML. This might need to be added as well and maybe you can point me to where this is done there.
- This behaviour might be undesired for some, but not for all use cases. Would you rather have this update functionality be opt-in with an env var?

Let me know what you think of this. We are maintaining our own fork of Directus anyway, so we don't push for this to be included, but I'd be happy to make changes to suit your requirements.

### PR DESCRIPTION
**Pull request by dasantonym**
## Scope

What's changed:

- Changed the handling of the updated user payload in the driver files under `api/auth/drivers/` for LDAP, OpenID and OAuth2. The name and email fields for users now get reset to the provider's payload on each login.

## Potential Risks / Drawbacks

- Might conflict with existing expectations for behaviour on login

## Review Notes / Questions

As already stated in the issue:

- The auth_data prop that is currently being updated is being set with the non-falsy value or null which I understand to be used to delete an existing value, correct? I have not used this for the names, but rather pass the value as is (undefined if not existing on the user info) so that the existing name remains if it is not supplied. Please correct me, if that is not the case.
- I have only made the changes for the above-mentioned files because I did not understand how this is handled in SAML. This might need to be added as well and maybe you can point me to where this is done there.
- This behaviour might be undesired for some, but not for all use cases. Would you rather have this update functionality be opt-in with an env var?

---

Fixes #20362

### CODE REVIEW

**Comment by dasantonym** [comment]
Hey @aidenfoxx, thanks for the PR! I have merged the latest from `main`.

One last question though: Is there a case where the `auth_data` would be undefined, but the name and email would be defined and then `auth_data` is set to `undefined` in the update? And would that even be a problem? I suppose unless you set the value to `null`, it does not get changed in the DB, right?

**Comment by aidenfoxx** [comment]
It's fine if "auth_data" is undefined in the update since it will be ignored, as you say, which is expected.😄

**Comment by dasantonym** [comment]
Ok, cool! Then, a newbie question: how do I proceed now? Not sure if I have to add a changeset to the API module since this is adding a new feature of sorts.

**Comment by dasantonym** [comment]
Hey @rijkvanzanten and @aidenfoxx, just a quick question if this PR can be merged or if you want me to do anything else? I'd like to merge soon it so I don't have to keep the PR up to date. Thanks!

**Comment by dasantonym** [comment]
No problem, thanks for getting back to me! You're right, I'll make the necessary changes and report back.

**Comment by madc** [comment]
This seems like a very helpful PR IMGO. Is there any movement on this?

**Comment by dasantonym** [comment]
Sorry for the standstill! I didn't finish this at the end of the year, and the new one has already started quite hectically. I hope to get this sorted out within the next two weeks.

**Comment by madc** [comment]
Thanks for the quick update and all the invested time. Greatly appreciated!

**Comment by dasantonym** [comment]
Hi @aidenfoxx & @rijkvanzanten, I have finally found the time to make the requested changes. Hope this is fine to merge now!

**Comment by kulogix** [comment]
For anyone looking to implement this functionality sooner, you can add the following hook (e.g., /extensions/hooks/auth-update-userinfo/index.js):

```javascript
export default async ({ filter }, { logger, database, services, getSchema }) => {
	const { UsersService } = services;
	const users = new UsersService({ schema: await getSchema(), knex: database });
	
	filter('auth.update', async (input, { identifier, provider, providerPayload }) => {
		if (providerPayload.userInfo && providerPayload.userInfo.sub) {
			logger.info(`Login for user with external identifier: ${identifier}`);
			// logger.info(`userPayload.auth_data: ${JSON.stringify(input)}`);
			// logger.info(`providerPayload: ${JSON.stringify(providerPayload)}`);
			const user = await database
				.select('id', 'first_name', 'last_name', 'email')
				.from('directus_users')
				.whereRaw('LOWER(??) = ?', ['external_identifier', identifier.toLowerCase()])
				.first();
			// logger.info(`Existing User: ${JSON.stringify(user)}`);
			
			if (
				user
				&& (providerPayload.userInfo.email && providerPayload.userInfo.given_name && providerPayload.userInfo.family_name)
				&& (
					providerPayload.userInfo.email != user.email
					|| providerPayload.userInfo.given_name != user.first_name
					|| providerPayload.userInfo.family_name != user.last_name
				)
			) {
				const userIds = await users.updateByQuery(
					{ filter: { id: user.id } }, 
					{ 
						email: providerPayload.userInfo.email,
						first_name: providerPayload.userInfo.given_name, 
						last_name: providerPayload.userInfo.family_name
					}
				);
				logger.info(`During login for user with external identifier ${identifier}, updated user ${user.id}: email from '${user.email}' to '${providerPayload.userInfo.email}', first name from '${user.first_name}' to '${providerPayload.userInfo.given_name}', last name from '${user.last_name}' to '${providerPayload.userInfo.family_name}'.`);
			} else if (
				user
				&& providerPayload.userInfo.email
				&& providerPayload.userInfo.email != user.email
			) {
				const userIds = await users.updateByQuery(
					{ filter: { id: user.id } },
					{
						email: providerPayload.userInfo.email
					}
				);
				logger.info(`During login for user with external identifier ${identifier}, updated user ${user.id}: email from '${user.email}' to '${providerPayload.userInfo.email}'.`);
			}
		}
		
		// Return the payload unchanged
		return input;
	});
};
```

File: api/src/auth/drivers/oauth2.ts [hunk 1]
```diff
@@ -176,9 +176,24 @@ export class OAuth2AuthDriver extends LocalAuthDriver {
...
+					auth_data: userPayload.auth_data,
+					first_name: userPayload.first_name,
+					last_name: userPayload.last_name,
+					email: userPayload.email,
+				};
+			} else {
+				emitPayload = {
+					auth_data: userPayload.auth_data,
+				};
+			}
```
Comment (aidenfoxx) [inline api/src/auth/drivers/oauth2.ts:192]: **nitpick:**
```suggestion
			let emitPayload = {
				auth_data: userPayload.auth_data,
			};

			if (syncUserInfo) {
				emitPayload = {
					...emitPayload,
					first_name: userPayload.first_name,
					last_name: userPayload.last_name,
					email: userPayload.email,
				};
			}
```

File: api/src/auth/drivers/openid.ts [hunk 2]
```diff
@@ -203,9 +203,24 @@ export class OpenIDAuthDriver extends LocalAuthDriver {
...
+					auth_data: userPayload.auth_data,
+					first_name: userPayload.first_name,
+					last_name: userPayload.last_name,
+					email: userPayload.email,
+				};
+			} else {
+				emitPayload = {
+					auth_data: userPayload.auth_data,
+				};
+			}
```
Comment (aidenfoxx) [inline api/src/auth/drivers/openid.ts:219]: **nitpick:**
```suggestion
			let emitPayload = {
				auth_data: userPayload.auth_data,
			};

			if (syncUserInfo) {
				emitPayload = {
					...emitPayload,
					first_name: userPayload.first_name,
					last_name: userPayload.last_name,
					email: userPayload.email,
				};
			}
```

**Review by aidenfoxx — APPROVED** [review]

**Comment by dasantonym** [comment]
Hey @aidenfoxx, thanks for reviewing the changes! I updated everything, but now the tests throw:

```
  Error: src/auth/drivers/oauth2.ts(188,6): error TS2353: Object literal may only specify known properties, and 'first_name' does not exist in type '{ auth_data: string | undefined; }'.
  Error: src/auth/drivers/openid.ts(215,6): error TS2353: Object literal may only specify known properties, and 'first_name' does not exist in type '{ auth_data: string | undefined; }'.
```

Definitely not a Typescript pro here, but I vaguely remember encountering this once when I started out on this PR and thought I had fixed it. Do you have an idea what to do here?

**Comment by dasantonym** [comment]
Ah, now I remember. This is why I used the clumsy if/else for setting the auth data. I had the exact same code you suggested in an earlier commit, but if you initially set the `emitPayload` object with only the one prop, you are not allowed to add more props later...

**Comment by dasantonym** [comment]
The latter suggestion did the trick, thanks a lot!

**Review by br41nslug — APPROVED** [review]

**Comment by br41nslug** [comment]
Due to lacking edit permissions on the PR to resolve the merge conflicts and add a changeset 😬 Moved this PR to #23595

**Pull request closed without merging**